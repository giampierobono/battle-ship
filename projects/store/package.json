{
  "name": "store",
  "version": "0.0.1",
  "dependencies": {
    "tslib": "^2.3.0"
  },
  "peerDependencies": {
    "@angular/common": "^12.2.0",
    "@angular/core": "^12.2.0",
    "@ngrx/effects": "^12.4.0",
    "@ngrx/store": "^12.4.0",
    "@smart/models": "^0.0.1",
    "@smart/store-tools": "0.0.1",
    "ramda": "^0.27.1",
    "rxjs": "^7.3.1"
  }
}
