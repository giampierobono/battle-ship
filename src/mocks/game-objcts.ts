import { BoatModel } from '../models';

export const mockGameSettings = {
  boardSize: 8,
  boatsPerPlayer: 4,
  numOfPlayers: 2,
};

export const mockBoatShapes = {
  '0': {
    shapeType: 0,
    shapeIndex: 0,
    segments: [
      {
        rowsCount: 2,
        colsCount: 1,
      },
      {
        rowsCount: 1,
        colsCount: 2,
      },
    ],
    shapeSize: 4,
  },
  '1': {
    shapeType: 1,
    shapeIndex: 0,
    segments: [
      {
        rowsCount: 2,
        colsCount: 2,
      },
    ],
    shapeSize: 4,
  },
  '2': {
    shapeType: 2,
    shapeIndex: 0,
    segments: [
      {
        rowsCount: 1,
        colsCount: 4,
      },
    ],
    shapeSize: 4,
  },
};

export const mockAllBoats: Array<Array<BoatModel>> = [
  [
    {
      positions: {
        '2': {
          '5': 'BOAT',
        },
        '3': {
          '5': 'HIT',
        },
        '4': {
          '5': 'BOAT',
          '6': 'BOAT',
        },
      },
      playerIndex: 0,
      boatIndex: 0,
      shapeKey: 0,
      hits: 1,
    },
    {
      positions: {
        '0': {
          '6': 'BOAT',
        },
        '1': {
          '6': 'BOAT',
        },
        '2': {
          '6': 'BOAT',
          '7': 'BOAT',
        },
      } as any,
      playerIndex: 0,
      boatIndex: 1,
      shapeKey: 0,
      hits: 0,
    },
    {
      positions: {
        '0': {
          '1': 'BOAT',
          '2': 'BOAT',
        },
        '1': {
          '1': 'BOAT',
          '2': 'BOAT',
        },
      } as any,
      playerIndex: 0,
      boatIndex: 2,
      shapeKey: 1,
      hits: 0,
    },
    {
      positions: {
        '4': {
          '0': 'BOAT',
          '1': 'BOAT',
        },
        '5': {
          '0': 'BOAT',
          '1': 'BOAT',
        },
      } as any,
      playerIndex: 0,
      boatIndex: 3,
      shapeKey: 1,
      hits: 0,
    },
  ],
  [
    {
      positions: {
        '6': {
          '2': 'BOAT',
          '3': 'BOAT',
          '4': 'BOAT',
          '5': 'BOAT',
        },
      } as any,
      playerIndex: 1,
      boatIndex: 0,
      shapeKey: 2,
      hits: 0,
    },
    {
      positions: {
        '2': {
          '0': 'BOAT',
          '1': 'BOAT',
          '2': 'BOAT',
          '3': 'BOAT',
        },
      } as any,
      playerIndex: 1,
      boatIndex: 1,
      shapeKey: 2,
      hits: 0,
    },
    {
      positions: {
        '0': {
          '2': 'BOAT',
          '3': 'BOAT',
          '4': 'BOAT',
          '5': 'BOAT',
        },
      } as any,
      playerIndex: 1,
      boatIndex: 2,
      shapeKey: 2,
      hits: 0,
    },
    {
      positions: {
        '3': {
          '6': 'HIT',
        },
        '4': {
          '6': 'BOAT',
        },
        '5': {
          '6': 'BOAT',
          '7': 'BOAT',
        },
      } as any,
      playerIndex: 1,
      boatIndex: 3,
      shapeKey: 0,
      hits: 1,
    },
  ],
];
