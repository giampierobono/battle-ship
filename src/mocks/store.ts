export const mockState = {
  battleShipCore: {
    players: {
      players: [
        {
          playerIndex: 0,
        },
        {
          playerIndex: 1,
        },
      ],
    },
    boats: {
      boats: [
        [
          {
            positions: {
              '2': {
                '5': 'BOAT',
              },
              '3': {
                '5': 'HIT',
              },
              '4': {
                '5': 'BOAT',
                '6': 'BOAT',
              },
            },
            playerIndex: 0,
            boatIndex: 0,
            shapeKey: 0,
            hits: 1,
          },
          {
            positions: {
              '0': {
                '6': 'HIT',
              },
              '1': {
                '6': 'BOAT',
              },
              '2': {
                '6': 'BOAT',
                '7': 'HIT',
              },
            },
            playerIndex: 0,
            boatIndex: 1,
            shapeKey: 0,
            hits: 2,
          },
          {
            positions: {
              '0': {
                '1': 'BOAT',
                '2': 'BOAT',
              },
              '1': {
                '1': 'BOAT',
                '2': 'BOAT',
              },
            },
            playerIndex: 0,
            boatIndex: 2,
            shapeKey: 1,
            hits: 0,
          },
          {
            positions: {
              '4': {
                '0': 'BOAT',
                '1': 'BOAT',
              },
              '5': {
                '0': 'BOAT',
                '1': 'BOAT',
              },
            },
            playerIndex: 0,
            boatIndex: 3,
            shapeKey: 1,
            hits: 0,
          },
        ],
        [
          {
            positions: {
              '6': {
                '2': 'BOAT',
                '3': 'BOAT',
                '4': 'BOAT',
                '5': 'BOAT',
              },
            },
            playerIndex: 1,
            boatIndex: 0,
            shapeKey: 2,
            hits: 0,
          },
          {
            positions: {
              '2': {
                '0': 'BOAT',
                '1': 'BOAT',
                '2': 'BOAT',
                '3': 'BOAT',
              },
            },
            playerIndex: 1,
            boatIndex: 1,
            shapeKey: 2,
            hits: 0,
          },
          {
            positions: {
              '0': {
                '2': 'BOAT',
                '3': 'BOAT',
                '4': 'BOAT',
                '5': 'BOAT',
              },
            },
            playerIndex: 1,
            boatIndex: 2,
            shapeKey: 2,
            hits: 0,
          },
          {
            positions: {
              '3': {
                '6': 'HIT',
              },
              '4': {
                '6': 'BOAT',
              },
              '5': {
                '6': 'BOAT',
                '7': 'BOAT',
              },
            },
            playerIndex: 1,
            boatIndex: 3,
            shapeKey: 0,
            hits: 1,
          },
        ],
      ],
    },
    boatShapes: {
      shapes: {
        '0': {
          shapeType: 0,
          shapeIndex: 0,
          segments: [
            {
              rowsCount: 2,
              colsCount: 1,
            },
            {
              rowsCount: 1,
              colsCount: 2,
            },
          ],
          shapeSize: 4,
        },
        '1': {
          shapeType: 1,
          shapeIndex: 0,
          segments: [
            {
              rowsCount: 2,
              colsCount: 2,
            },
          ],
          shapeSize: 4,
        },
        '2': {
          shapeType: 2,
          shapeIndex: 0,
          segments: [
            {
              rowsCount: 1,
              colsCount: 4,
            },
          ],
          shapeSize: 4,
        },
      },
    },
    settings: {
      boardSize: 8,
      boatsPerPlayer: 4,
      numOfPlayers: 2,
    },
    game: {
      playerTurn: 1,
      isCurrentTurnOver: true,
      isGameOver: false,
      playersMoves: {
        '0': {
          '1': {
            '7': 'MISS',
          },
          '2': {
            '5': 'MISS',
          },
          '3': {
            '3': 'MISS',
            '6': 'HIT',
          },
          '5': {
            '5': 'MISS',
          },
        },
        '1': {
          '0': {
            '6': 'HIT',
          },
          '2': {
            '7': 'HIT',
          },
          '3': {
            '5': 'HIT',
          },
          '5': {
            '4': 'MISS',
            '6': 'MISS',
          },
        },
      },
      playersSunkBoats: {},
    },
  },
};
