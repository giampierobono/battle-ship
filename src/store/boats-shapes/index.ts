export * from './boat-shapes.actions';
export * from './boat-shapes.reducer';
export * from './boat-shape.selectors';
export * from './models';
