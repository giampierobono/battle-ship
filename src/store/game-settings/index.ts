export * from './game-settings.actions';
export * from './game-settings.effects';
export * from './game-settings.reducer';
export * from './game-settings.selectors';
export * from './models';
