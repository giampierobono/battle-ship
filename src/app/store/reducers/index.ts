import { ActionReducer, ActionReducerMap, INIT, MetaReducer } from '@ngrx/store';
import { localStorageSync } from 'ngrx-store-localstorage';
import { resetGame } from '../actions/reset-game.actions';

const localStorageSyncReducer = (reducer: ActionReducer<AppState>): ActionReducer<AppState> =>
  localStorageSync({
    keys: [
      {
        battleShipCore: {
          encrypt: (state) => btoa(state),
          decrypt: (state) => atob(state),
        },
      },
    ],
    rehydrate: true,
  })(reducer);

const resetGameReducer = (reducer: ActionReducer<AppState>): ActionReducer<AppState> => {
  return (state, action) => {
    if (action != null && action.type === resetGame.type) {
      localStorage.removeItem('battleShipCore');
      return reducer(undefined, { type: INIT });
    }
    return reducer(state, action);
  };
};

export interface AppState {}

export const reducers: ActionReducerMap<AppState> = {};

export const metaReducers: MetaReducer<AppState>[] = [resetGameReducer, localStorageSyncReducer];
