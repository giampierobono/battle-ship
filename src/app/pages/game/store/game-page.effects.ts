import { Injectable } from '@angular/core';
import { Actions, createEffect, ofType } from '@ngrx/effects';
import { Store } from '@ngrx/store';
import { AppState } from '../../../store/reducers';
import { Router } from '@angular/router';
import { tap } from 'rxjs/operators';
import { resetGame } from '../../../store/actions/reset-game.actions';

@Injectable()
export class GamePageEffects {
  public resetGame$ = createEffect(
    () =>
      this.actions$.pipe(
        ofType(resetGame),
        tap(() => localStorage.removeItem('')),
        tap(() => this.router.navigate(['lobby']))
      ),
    { dispatch: false }
  );

  constructor(private actions$: Actions, private store: Store<AppState>, private router: Router) {}
}
