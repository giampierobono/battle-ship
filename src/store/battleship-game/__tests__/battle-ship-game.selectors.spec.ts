import {
  getAllBoatsPositionsForCurrentPlayer,
  getBattleShipGameState,
  getBoatForCurrentPlayer,
  getCurrentPlayerSunkBoats,
  getIsCurrentTurnOver,
  getIsGameOver,
  getPlayerMoves,
  getPlayersSunkBoats,
  getPlayerTurn,
  getSunkBoatPerPlayer,
  getTotalFiredShots,
} from '../battle-ship-game.selectors';
import { mockState } from '../../../mocks/store';

describe('Battleship game selectors', () => {
  describe('getBattleShipGameState', () => {
    it('should return game state', () => {
      expect(getBattleShipGameState(mockState)).toEqual(mockState.battleShipCore.game);
    });
  });

  describe('getPlayerTurn', () => {
    it('should return current player turn', () => {
      expect(getPlayerTurn(mockState)).toBe(1);
    });
  });

  describe('getPlayersSunkBoats', () => {
    it('should return sunm boats per player', () => {
      expect(getPlayersSunkBoats(mockState)).toMatchSnapshot();
    });
  });

  describe('getIsGameOver', () => {
    it('should return isGameOver property from state', () => {
      expect(getIsGameOver(mockState)).toBeFalsy();
    });
  });

  describe('getCurrentPlayerSunkBoats', () => {
    it('should return current player sunk boats property from state', () => {
      expect(getCurrentPlayerSunkBoats(mockState)).toMatchSnapshot();
    });
  });

  describe('getIsCurrentTurnOver', () => {
    it('should return isCurrentTurnOver property from state', () => {
      expect(getIsCurrentTurnOver(mockState)).toBeTruthy();
    });
  });

  describe('getPlayerMoves', () => {
    it('should return current player moves property from state', () => {
      expect(getPlayerMoves(mockState)).toMatchSnapshot();
    });
  });

  describe('getTotalFiredShots', () => {
    it('should return total fired shots sum', () => {
      expect(getTotalFiredShots(mockState)).toBe(10);
    });
  });

  describe('getSunkBoatPerPlayer', () => {
    it('should return sunk boats per player', () => {
      expect(getSunkBoatPerPlayer(mockState)).toMatchSnapshot();
    });
  });

  describe('getBoatForCurrentPlayer', () => {
    it('should return current player boats', () => {
      expect(getBoatForCurrentPlayer(mockState)).toMatchSnapshot();
    });
  });

  describe('getAllBoatsPositionsForCurrentPlayer', () => {
    it('should return all boats positions for current player', () => {
      expect(getAllBoatsPositionsForCurrentPlayer(mockState)).toMatchSnapshot();
    });
  });
});
