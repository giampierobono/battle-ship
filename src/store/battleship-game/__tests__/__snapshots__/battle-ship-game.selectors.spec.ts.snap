// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Battleship game selectors getAllBoatsPositionsForCurrentPlayer should return all boats positions for current player 1`] = `
Object {
  "0": Object {
    "2": "BOAT",
    "3": "BOAT",
    "4": "BOAT",
    "5": "BOAT",
  },
  "2": Object {
    "0": "BOAT",
    "1": "BOAT",
    "2": "BOAT",
    "3": "BOAT",
  },
  "3": Object {
    "6": "HIT",
  },
  "4": Object {
    "6": "BOAT",
  },
  "5": Object {
    "6": "BOAT",
    "7": "BOAT",
  },
  "6": Object {
    "2": "BOAT",
    "3": "BOAT",
    "4": "BOAT",
    "5": "BOAT",
  },
}
`;

exports[`Battleship game selectors getBoatForCurrentPlayer should return current player boats 1`] = `
Array [
  Object {
    "boatIndex": 0,
    "hits": 0,
    "playerIndex": 1,
    "positions": Object {
      "6": Object {
        "2": "BOAT",
        "3": "BOAT",
        "4": "BOAT",
        "5": "BOAT",
      },
    },
    "shapeKey": 2,
  },
  Object {
    "boatIndex": 1,
    "hits": 0,
    "playerIndex": 1,
    "positions": Object {
      "2": Object {
        "0": "BOAT",
        "1": "BOAT",
        "2": "BOAT",
        "3": "BOAT",
      },
    },
    "shapeKey": 2,
  },
  Object {
    "boatIndex": 2,
    "hits": 0,
    "playerIndex": 1,
    "positions": Object {
      "0": Object {
        "2": "BOAT",
        "3": "BOAT",
        "4": "BOAT",
        "5": "BOAT",
      },
    },
    "shapeKey": 2,
  },
  Object {
    "boatIndex": 3,
    "hits": 1,
    "playerIndex": 1,
    "positions": Object {
      "3": Object {
        "6": "HIT",
      },
      "4": Object {
        "6": "BOAT",
      },
      "5": Object {
        "6": "BOAT",
        "7": "BOAT",
      },
    },
    "shapeKey": 0,
  },
]
`;

exports[`Battleship game selectors getCurrentPlayerSunkBoats should return current player sunk boats property from state 1`] = `Array []`;

exports[`Battleship game selectors getPlayerMoves should return current player moves property from state 1`] = `
Object {
  "0": Object {
    "6": "HIT",
  },
  "2": Object {
    "7": "HIT",
  },
  "3": Object {
    "5": "HIT",
  },
  "5": Object {
    "4": "MISS",
    "6": "MISS",
  },
}
`;

exports[`Battleship game selectors getPlayersSunkBoats should return sunm boats per player 1`] = `Object {}`;

exports[`Battleship game selectors getSunkBoatPerPlayer should return sunk boats per player 1`] = `
Array [
  0,
  0,
]
`;
